Move RecordFunction into ATen

ilia-cher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #37587 RecordFunction in Dispatcher
* #37639 Use TensorMethods.cpp
* #37775 Memory profiling
* #37774 Move ThreadLocalDebugInfo to c10
* #36291 Make profiler thread local
* #37745 Propagate TLS callbacks with ThreadLocalState
* **#37548 Move RecordFunction into ATen**
* #37491 Make RecordFunction callbacks thread local and modernize interface

Summary:
Moving RecordFunction from torch::autograd::profiler into at namespace

Test Plan:
CI

Differential Revision: [D21315852](https://our.internmc.facebook.com/intern/diff/D21315852)

Commit:
Move RecordFunction into ATen

Summary:
Moving RecordFunction from torch::autograd::profiler into at namespace

Test Plan:
CI

[ghstack-poisoned]

Commit:
Update on "Move RecordFunction into ATen"


Summary:
Moving RecordFunction from torch::autograd::profiler into at namespace

Test Plan:
CI

[ghstack-poisoned]

Commit:
Update on "Move RecordFunction into ATen"


Summary:
Moving RecordFunction from torch::autograd::profiler into at namespace

Test Plan:
CI

Differential Revision: [D21315852](https://our.internmc.facebook.com/intern/diff/D21315852)

[ghstack-poisoned]

Commit:
Update on "Move RecordFunction into ATen"


Summary:
Moving RecordFunction from torch::autograd::profiler into at namespace

Test Plan:
CI

Differential Revision: [D21315852](https://our.internmc.facebook.com/intern/diff/D21315852)

[ghstack-poisoned]

Commit:
Update on "Move RecordFunction into ATen"


Summary:
Moving RecordFunction from torch::autograd::profiler into at namespace

Test Plan:
CI

Differential Revision: [D21315852](https://our.internmc.facebook.com/intern/diff/D21315852)

[ghstack-poisoned]

Commit:
Update on "Move RecordFunction into ATen"


Summary:
Moving RecordFunction from torch::autograd::profiler into at namespace

Test Plan:
CI

Differential Revision: [D21315852](https://our.internmc.facebook.com/intern/diff/D21315852)

[ghstack-poisoned]

Commit:
Update on "Move RecordFunction into ATen"


Summary:
Moving RecordFunction from torch::autograd::profiler into at namespace

Test Plan:
CI

Differential Revision: [D21315852](https://our.internmc.facebook.com/intern/diff/D21315852)

[ghstack-poisoned]

Commit:
Update on "Move RecordFunction into ATen"


Summary:
Moving RecordFunction from torch::autograd::profiler into at namespace

Test Plan:
CI

Differential Revision: [D21315852](https://our.internmc.facebook.com/intern/diff/D21315852)

[ghstack-poisoned]

Commit:
Update on "Move RecordFunction into ATen"


Summary:
Moving RecordFunction from torch::autograd::profiler into at namespace

Test Plan:
CI

Differential Revision: [D21315852](https://our.internmc.facebook.com/intern/diff/D21315852)

[ghstack-poisoned]

Commit:
Update on "Move RecordFunction into ATen"


Summary:
Moving RecordFunction from torch::autograd::profiler into at namespace

Test Plan:
CI

Differential Revision: [D21315852](https://our.internmc.facebook.com/intern/diff/D21315852)

[ghstack-poisoned]

Commit:
Update on "Move RecordFunction into ATen"


Summary:
Moving RecordFunction from torch::autograd::profiler into at namespace

Test Plan:
CI

Differential Revision: [D21315852](https://our.internmc.facebook.com/intern/diff/D21315852)

[ghstack-poisoned]

Commit:
Update on "Move RecordFunction into ATen"


Summary:
Moving RecordFunction from torch::autograd::profiler into at namespace

Test Plan:
CI

Differential Revision: [D21315852](https://our.internmc.facebook.com/intern/diff/D21315852)

[ghstack-poisoned]

Commit:
Update on "Move RecordFunction into ATen"


Summary:
Moving RecordFunction from torch::autograd::profiler into at namespace

Test Plan:
CI

Differential Revision: [D21315852](https://our.internmc.facebook.com/intern/diff/D21315852)

[ghstack-poisoned]

Commit:
Update on "Move RecordFunction into ATen"


Summary:
Moving RecordFunction from torch::autograd::profiler into at namespace

Test Plan:
CI

Differential Revision: [D21315852](https://our.internmc.facebook.com/intern/diff/D21315852)

[ghstack-poisoned]

Commit:
Update on "Move RecordFunction into ATen"


Summary:
Moving RecordFunction from torch::autograd::profiler into at namespace

Test Plan:
CI

Differential Revision: [D21315852](https://our.internmc.facebook.com/intern/diff/D21315852)

[ghstack-poisoned]

dzhulgakov commented on docs/source/notes/large_scale_deployments.rst:
```diff
@@ -27,7 +27,7 @@ across the entire set of machines.
 New callbacks for any operator invocation can be added with
 ``torch::autograd::profiler::addGlobalCallback``. Hooks will be called with
-``torch::autograd::profiler::RecordFunction`` struct that describes invocation
+``at::RecordFunction`` struct that describes invocation
```
for public C++ API I think we're trying to use `torch::` namespace. So maybe add `using` somewhere (if we haven't already) and refer to it at `torch::RecordFunction` in the docs?

Commit:
Update on "Move RecordFunction into ATen"


Summary:
Moving RecordFunction from torch::autograd::profiler into at namespace

Test Plan:
CI

Differential Revision: [D21315852](https://our.internmc.facebook.com/intern/diff/D21315852)

[ghstack-poisoned]

Commit:
Update on "Move RecordFunction into ATen"


Summary:
Moving RecordFunction from torch::autograd::profiler into at namespace

Test Plan:
CI

Differential Revision: [D21315852](https://our.internmc.facebook.com/intern/diff/D21315852)

[ghstack-poisoned]

Commit:
Update on "Move RecordFunction into ATen"


Summary:
Moving RecordFunction from torch::autograd::profiler into at namespace

Test Plan:
CI

Differential Revision: [D21315852](https://our.internmc.facebook.com/intern/diff/D21315852)

[ghstack-poisoned]

Commit:
Update on "Move RecordFunction into ATen"


Summary:
Moving RecordFunction from torch::autograd::profiler into at namespace

Test Plan:
CI

Differential Revision: [D21315852](https://our.internmc.facebook.com/intern/diff/D21315852)

[ghstack-poisoned]

Commit:
Update on "Move RecordFunction into ATen"


Summary:
Moving RecordFunction from torch::autograd::profiler into at namespace

Test Plan:
CI

Differential Revision: [D21315852](https://our.internmc.facebook.com/intern/diff/D21315852)

[ghstack-poisoned]

Commit:
Update on "Move RecordFunction into ATen"


Summary:
Moving RecordFunction from torch::autograd::profiler into at namespace

Test Plan:
CI

Differential Revision: [D21315852](https://our.internmc.facebook.com/intern/diff/D21315852)

[ghstack-poisoned]

Commit:
Update on "Move RecordFunction into ATen"


Summary:
Moving RecordFunction from torch::autograd::profiler into at namespace

Test Plan:
CI

Differential Revision: [D21315852](https://our.internmc.facebook.com/intern/diff/D21315852)

[ghstack-poisoned]

Commit:
Update on "Move RecordFunction into ATen"


Summary:
Moving RecordFunction from torch::autograd::profiler into at namespace

Test Plan:
CI

Differential Revision: [D21315852](https://our.internmc.facebook.com/intern/diff/D21315852)

[ghstack-poisoned]

Commit:
Update on "Move RecordFunction into ATen"


Summary:
Moving RecordFunction from torch::autograd::profiler into at namespace

Test Plan:
CI

Differential Revision: [D21315852](https://our.internmc.facebook.com/intern/diff/D21315852)

[ghstack-poisoned]

Commit:
Update on "Move RecordFunction into ATen"


Summary:
Moving RecordFunction from torch::autograd::profiler into at namespace

Test Plan:
CI

Differential Revision: [D21315852](https://our.internmc.facebook.com/intern/diff/D21315852)

[ghstack-poisoned]

Commit:
Update on "Move RecordFunction into ATen"


Summary:
Moving RecordFunction from torch::autograd::profiler into at namespace

Test Plan:
CI

Differential Revision: [D21315852](https://our.internmc.facebook.com/intern/diff/D21315852)

[ghstack-poisoned]

Commit:
Update on "Move RecordFunction into ATen"


Summary:
Moving RecordFunction from torch::autograd::profiler into at namespace

Test Plan:
CI

Differential Revision: [D21315852](https://our.internmc.facebook.com/intern/diff/D21315852)

[ghstack-poisoned]

Commit:
Update on "Move RecordFunction into ATen"


Summary:
Moving RecordFunction from torch::autograd::profiler into at namespace

Test Plan:
CI

Differential Revision: [D21315852](https://our.internmc.facebook.com/intern/diff/D21315852)

[ghstack-poisoned]

Commit:
Update on "Move RecordFunction into ATen"


Summary:
Moving RecordFunction from torch::autograd::profiler into at namespace

Test Plan:
CI

Differential Revision: [D21315852](https://our.internmc.facebook.com/intern/diff/D21315852)

[ghstack-poisoned]

Commit:
Update on "Move RecordFunction into ATen"


Summary:
Moving RecordFunction from torch::autograd::profiler into at namespace

Test Plan:
CI

Differential Revision: [D21315852](https://our.internmc.facebook.com/intern/diff/D21315852)

[ghstack-poisoned]

Commit:
Update on "Move RecordFunction into ATen"


Summary:
Moving RecordFunction from torch::autograd::profiler into at namespace

Test Plan:
CI

Differential Revision: [D21315852](https://our.internmc.facebook.com/intern/diff/D21315852)

[ghstack-poisoned]

Commit:
Update on "Move RecordFunction into ATen"


Summary:
Moving RecordFunction from torch::autograd::profiler into at namespace

Test Plan:
CI

Differential Revision: [D21315852](https://our.internmc.facebook.com/intern/diff/D21315852)

[ghstack-poisoned]

Commit:
Update on "Move RecordFunction into ATen"


Summary:
Moving RecordFunction from torch::autograd::profiler into at namespace

Test Plan:
CI

Differential Revision: [D21315852](https://our.internmc.facebook.com/intern/diff/D21315852)

[ghstack-poisoned]

Commit:
Update on "Move RecordFunction into ATen"


Summary:
Moving RecordFunction from torch::autograd::profiler into at namespace

Test Plan:
CI

Differential Revision: [D21315852](https://our.internmc.facebook.com/intern/diff/D21315852)

[ghstack-poisoned]

Commit:
Update on "Move RecordFunction into ATen"


Summary:
Moving RecordFunction from torch::autograd::profiler into at namespace

Test Plan:
CI

Differential Revision: [D21315852](https://our.internmc.facebook.com/intern/diff/D21315852)

[ghstack-poisoned]

Commit:
Update on "Move RecordFunction into ATen"


Summary:
Moving RecordFunction from torch::autograd::profiler into at namespace

Test Plan:
CI

Differential Revision: [D21315852](https://our.internmc.facebook.com/intern/diff/D21315852)

[ghstack-poisoned]
